python/dglke/models/pytorch/ke_tensor.py

```diff
@@ -0,0 +1,308 @@
+# -*- coding: utf-8 -*-
+#
+# tensor_models.py
+#
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+KG Sparse embedding
+"""
+import os
+import numpy as np
+
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as functional
+import torch.nn.init as INIT
+
+import torch.multiprocessing as mp
+from torch.multiprocessing import Queue
+from _thread import start_new_thread
+import traceback
+from functools import wraps
+
+from .. import *
+
+from .tensor_models import thread_wrapped_func
+
+@thread_wrapped_func
+def async_update(num_thread, emb, queue):
+    """Asynchronous embedding update for entity embeddings.
+    How it works:
+        1. trainer process push entity embedding update requests into the queue.
+        2. async_update process pull requests from the queue, calculate
+           the gradient state and gradient and write it into entity embeddings.
+
+    Parameters
+    ----------
+    args :
+        Global confis.
+    emb : ExternalEmbedding
+        The entity embeddings.
+    queue:
+        The request queue.
+    """
+    th.set_num_threads(num_thread)
+    while True:
+        (grad_indices, grad_values, gpu_id) = queue.get()
+        clr = emb.lr
+        if grad_indices is None:
+            return
+        with th.no_grad():
+            grad_sum = (grad_values * grad_values).mean(1)
+            device = emb.state_sum.device
+            if device != grad_indices.device:
+                grad_indices = grad_indices.to(device)
+            if device != grad_sum.device:
+                grad_sum = grad_sum.to(device)
+
+            emb.state_sum.index_add_(0, grad_indices, grad_sum)
+            std = emb.state_sum[grad_indices]  # _sparse_mask
+            if gpu_id >= 0:
+                std = std.cuda(gpu_id)
+            std_values = std.sqrt_().add_(1e-10).unsqueeze(1)
+            tmp = (-clr * grad_values / std_values)
+            if tmp.device != device:
+                tmp = tmp.to(device)
+            emb.emb.index_add_(0, grad_indices, tmp)
+
+
+class KGEmbedding:
```

Comment:
where is it used for training?

Reply:
Currently only the load() method is used to load the embeddings.